File: neet_code_150_cpp/stack/daily_tmp.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <stack>

// Rustの模範解答より
class SolutionAnsRust {
    public:
        std::vector<int> daily_tmp(const std::vector<int>& temperatures) {
            auto len_tmp = temperatures.size();
            std::vector<int> result(len_tmp,0);
            std::stack<std::pair<int,std::size_t>> stack;

            for(std::size_t i = 0; i < len_tmp; ++i) {
                while(!stack.empty() && temperatures[i] > stack.top().first) {
                    auto stack_idx = stack.top().second;
                    stack.pop();
                    result[stack_idx] = static_cast<int>(i - stack_idx);
                }

                stack.push({temperatures[i], i});
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<int> dailyTemperatures(const std::vector<int>& temperatures) {
            int n = temperatures.size();

            // pair: [index, temp]
            std::stack<std::pair<int, int>> stk;
            std::vector<int> result(n);

            for (int i = 0; i < n; i++) {
                int currDay = i;
                int currTemp = temperatures[i];

                while (!stk.empty() && stk.top().second < currTemp) {
                    int prevDay = stk.top().first;
                    int prevTemp = stk.top().second;
                    stk.pop();

                    result[prevDay] = currDay - prevDay;
                }

                stk.push({currDay, currTemp});
            }

            return result;
        }
};

// 構造化束縛を使うとちょっとスッキリ
class SolutionRe {
    public:
        std::vector<int> dailyTemperatures(const std::vector<int>& temperatures) {
            auto n = temperatures.size();
            
            std::stack<std::pair<int, int>> stk;
            std::vector<int> result(n);

            for(int i = 0; i < n; ++i) {
                int currDay = i;
                int currTemp = temperatures[i];

                while(!stk.empty() && stk.top().second < currTemp) {
                    // 構造化束縛!
                    auto [prevDay, prevTemp] = stk.top();
                    stk.pop();

                    result[prevDay] = currDay - prevDay;
                }

                stk.push({currDay, currTemp});
            }

            return result;
        }
};

int main(void) {
    const auto case_1 = std::vector{73, 74, 75, 71, 69, 72, 76, 73};
    const auto case_2 = std::vector{30, 40, 50, 60};
    const auto case_3 = std::vector{30, 60, 90};

    SolutionAnsRust s_ans_rust;

    auto res_1_rust = s_ans_rust.daily_tmp(case_1);
    for(const auto& v: res_1_rust) {
        std::cout << v << " ";
    }
    std::cout << std::endl;

    auto res_2_rust = s_ans_rust.daily_tmp(case_2);
    for(const auto& v: res_2_rust) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
    auto res_3_rust = s_ans_rust.daily_tmp(case_3);
    for(const auto& v: res_3_rust) {
        std::cout << v << " ";
    }
    std::cout << std::endl;

    SolutionAns s_ans;
    auto res_1_ans = s_ans.dailyTemperatures(case_1);
    for(const auto& v: res_1_ans) {
        std::cout << v << " ";
    }
    std::cout << std::endl;

    auto res_2_ans = s_ans.dailyTemperatures(case_2);
    for(const auto& v: res_2_ans) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
    auto res_3_ans = s_ans.dailyTemperatures(case_3);
    for(const auto& v: res_3_ans) {
        std::cout << v << " ";
    }
    std::cout << std::endl;
}

File: neet_code_150/src/stack/daily_tmp.rs
struct Solution {}

impl Solution {
    // TLE
    pub fn daily_tmp_sq(temperatures: Vec<i32>) -> Vec<i32> {
        let mut res = vec![0; temperatures.len()];
        for i in 0..temperatures.len() {
            let tmp_v = temperatures[i];
            for j in i + 1..temperatures.len() {
                if tmp_v < temperatures[j] {
                    res[i] = (j - i) as i32;
                    break;
                }
            }
        }
        res
    }

    // 解けなかった😭
    pub fn daily_tmp(temperatures: Vec<i32>) -> Vec<i32> {
        let mut res = vec![0; temperatures.len()];
        let mut stack = vec![];

        for i in 0..temperatures.len() {
            stack.push(temperatures[i]);
        }

        res
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let mut res = vec![0; temperatures.len()];
        let mut stack: Vec<(i32, usize)> = vec![]; // (temp, index)

        for (i, val) in temperatures.iter().enumerate() {
            while !stack.is_empty() && *val > stack.last().unwrap().0 {
                let (_, stack_index) = stack.pop().unwrap();
                res[stack_index] = (i - stack_index) as i32;
            }

            stack.push((*val, i));
        }
        res
    }
}

fn main() {
    let case_1 = vec![73, 74, 75, 71, 69, 72, 76, 73]; // [1,1,4,2,1,1,0,0]
    let case_2 = vec![30, 40, 50, 60]; // [1,1,10]
    let case_3 = vec![30, 60, 90]; // [1,1,0]

    println!("case_1: {:?}", Solution::daily_tmp_sq(case_1.clone()));
    println!("case_2: {:?}", Solution::daily_tmp_sq(case_2.clone()));
    println!("case_3: {:?}", Solution::daily_tmp_sq(case_3.clone()));

    println!("case_1: {:?}", SolutionAns::daily_temperatures(case_1.clone()));
    println!("case_2: {:?}", SolutionAns::daily_temperatures(case_2.clone()));
    println!("case_3: {:?}", SolutionAns::daily_temperatures(case_3.clone()));
}
